numpy/core/numeric.py

```diff
@@ -2357,8 +2359,10 @@ def within_tol(x, y, atol, rtol):
         # lib.stride_tricks, though, so we can't import it here.
         x = x * ones_like(cond)
         y = y * ones_like(cond)
+        atol_f = np.broadcast_to(atol, x.shape)[finite] if atol.shape else atol
```

Comment:
So this is weird, above it is noted we are not allowed `broadcast_arrays`, but here we use `broadcast_to`. So, clearly that comment is out of date, which means we can just do (following the definition of `cond` - rest can go):
```
x, y, atol, rtol = np.broadcast_arrays(x, y, atol, rtol, subok=True)
cond[finite] = within_tol(x[finite], y[finite], atol[finite], rtol[finite])
```
(note addition of `subok=True` - useful for subclasses)

Could you check whether the tests still pass with this?

Reply:
Yup, this looks better to me. 

It turns out that in NumPy main, `isclose` is very flexible about the shape of `atol` and `rtol` when there are no NaNs. 

```python3
import numpy as np
np.isclose([1, 2], [[2], [1]], atol=[[[0.01]], [[2]]])
# array([[[False,  True],
#         [ True, False]],
#
#        [[ True,  True],
#         [ True,  True]]])
```

Before the commit, this would have failed if there were a NaN in one of the arrays because `atol` cannot be broadcast _to_ the shape of `x` (which is really the braodcasted shape of `x` and `y` at this point in the code).

Now this will work even when there are NaNs.

```python3
np.isclose([np.nan, 2], [[2], [np.nan]], atol=[[[0.01]], [[3]]], equal_nan=True)
# array([[[False,  True],
#         [ True, False]],
#
#        [[False,  True],
#         [ True, False]]])
```